### PR TITLE
feat!(x/gov): implement a minimum amount per deposit (backport #19312)

### PR DESCRIPTION
**Pull request by mattverse**
## Description

Backport: https://github.com/cosmos/cosmos-sdk/pull/19312 and https://github.com/cosmos/cosmos-sdk/pull/18146

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

## Summary by CodeRabbit

- **New Features**
	- Introduced a minimum deposit ratio for governance proposals to ensure deposits meet a specified proportion of the minimum required value.
- **Bug Fixes**
	- Adjusted deposit amounts in various test cases and simulations to comply with the new minimum deposit ratio requirements.
- **Tests**
	- Removed outdated tests and added new ones to validate deposit amounts and ratios.
	- Updated test scenarios to reflect changes in deposit validation logic.
- **Documentation**
	- Updated the CHANGELOG to include details about the new minimum deposit feature in the governance module.
- **Refactor**
	- Enhanced deposit validation logic across several modules to enforce the new minimum deposit ratio.
- **Chores**
	- Updated simulation and migration scripts to accommodate the new minimum deposit ratio parameter.

### CODE REVIEW

**Comment by czarcas7ic** [comment]
@mattverse can you ping when e2e test is passing and i will review

**Comment by mattverse** [comment]
@czarcas7ic this is r4r!

**Comment by czarcas7ic** [comment]
Dont we already have this here https://github.com/osmosis-labs/cosmos-sdk/pull/510/files#diff-08bbfe14de423426949f72887bae76f115db4bc2478cd7c3d60e44a4325c7711R221

**Comment by czarcas7ic** [comment]
Yeah I am fairly confident this is already implemented but just with a different name @mattverse, can you double check my claim

**Comment by czarcas7ic** [comment]
Going to set as draft until response

**Comment by mattverse** [comment]
@czarcas7ic This is different from the initial minimum deposit we have. 
The Initial minimum deposit state we have sets the min amount that needs to be inputted for the initial deposit. 

This state is for having a minimum deposit for each deposit tx user can make.

**Comment by mattverse** [comment]
@czarcas7ic another reason Im sure that these are two different things is that upstream also has these two different states. Linking the proto file below for reference. 

https://github.com/cosmos/cosmos-sdk/blob/main/x/gov/proto/cosmos/gov/v1/gov.proto#L296
https://github.com/cosmos/cosmos-sdk/blob/main/x/gov/proto/cosmos/gov/v1/gov.proto#L343

File: x/gov/keeper/deposit.go [hunk 1]
```diff
@@ -118,8 +119,54 @@ func (keeper Keeper) AddDeposit(ctx sdk.Context, proposalID uint64, depositorAdd
...
+	// v47 does not have expedited proposals so we always use params.MinDeposit
+	minDepositAmount := params.MinDeposit
+	minDepositRatio, err := sdk.NewDecFromStr(params.GetMinDepositRatio())
+	if err != nil {
+		return false, err
+	}
+
+	if err := keeper.validateDepositDenom(ctx, params, depositAmount); err != nil {
+		return false, err
+	}
```
Comment (czarcas7ic) [inline x/gov/keeper/deposit.go]: We do have expedited proposals, so we need to include the expedited proposal logic @mattverse

File: x/gov/simulation/operations.go [hunk 2]
```diff
@@ -478,7 +485,10 @@ func randomDeposit(
 	}
 	amount = amount.Add(minAmount)
 
-	if amount.GT(spendableBalance) {
+	// NOTE: backport from v50
+	amount = amount.MulRaw(3) // 3x what's required // TODO: this is a hack, we need to be able to calculate the correct amount using params
+
```
Comment (czarcas7ic) [inline x/gov/simulation/operations.go:490]: This doesn't make sense to me, but I assume its just backporting what they did

File: x/gov/keeper/deposit.go (hunk 1 above)
Comment (mattverse) [reply x/gov/keeper/deposit.go]: Oof good catch! Fixed in the most recent commit!

**Review by czarcas7ic — APPROVED** [review]

**Pull request merged**